fix #1109 (licence sur articles)

pierre-24 opened the pull request:
| Q | R |
| --- | --- |
| Correction de bugs ? | non |
| Nouvelle Fonctionnalité ? | oui |
| Tickets concernés | #1109 |

Ce qui a été fait :
- Ajouter un champ "licence" dans la base de donnée sur la table article (donc faire un `manage.py migrate`)
- Faire en sorte que ce champ licence se retrouve dans le manifest.json
- Ajout de la possibilité d'éditer la licence dans les formulaires correspondants.
- Visualisation de la licence dans les templates correspondants.

(le tout largement inspiré de ce qui se fait pour les tutos)

Mais aussi : 
- Déplacement de `template/article/view.html` vers `/templates/articles/user/view_online.html`, pour garder le même workflow que sur les tutoriaux
- ~~Retirer le `pager` de `view_online.html`, qui n'as rien à y faire, un article ne fera jamais plusieurs pages.~~ **je n'avais pas le droit de faire ça.**
# Note pour QA

Ce n'est pas absolument urgent, l'issue a le label "V1" et pas "RCx"

Il faut tester tout le _workflow_ des articles, donc : 
- **Migrer la base** avec `manage.py migrate`
- **Tenter de créer un nouvel article**, vérifier qu'on sais bien lui assigner une licence. **Faire une prévisualisation avant de créer l'article** et vérifier qu'on retombe bien sur le formulaire et que la licence reste bonne (ce n'était pas le cas avant) ;
- **Vérifier la présence de la licence** sur la preview en _off_ de l'article, à droite du titre
- **Modifier la licence** et vérifier que ça impacte bien l'affichage en _off_. **Regarder le manifest.JSON**, vérifier que la licence y est bien, et qu'elle change si on en change.
- **Valider l'article** et vérifier que la licence est présente sur la page en _online_.
- Si c'était possible, **regarder si le comportement d'anciens articles n'est pas affecté** par la mise à jour de la base. J'espère que non, mais je n'ais pas d'ancien articles pour le vérifier.

(et avant toute chose, [n'oubliez pas de créer des licences bidons sur votre version locale](http://zestedesavoir.com/forums/sujet/665/gestion-des-licences/), ou vous n'aurez pas grand chose à QA)

pierre-24 committed:
fix #1109 (licence sur articles)

Ce qui a été fait :
- Ajouter un champ "licence" dans la base de donnée sur la table
  article (donc faire un `manage.py migrate`)
- Faire en sorte que ce champ licence se retrouve dans le
  manifest.json
- Ajout de la possibilité d'éditer la licence dans les formulaires
  correspondants.
- Visualisation de la licence dans les templates correspondants.
- Déplacement de template/article/view.html vers /templates/
  articles/user/view_online.html, pour garder le même workflow
  que sur les tutoriaux
(le tout largement inspiré de ce qui se fait pour les tutos)

Eskimon commented:
Hop, Review faite de mon coté...
Tout à l'air bon SAUF:
- L'affichage de la licence casse un peu le design (mais mon _dist_ est ptet pas à jour non plus)
- Si je met une licence, je ne peux plus l'enlever

> regarder si le comportement d'anciens articles n'est pas affecté

Pas testé non plus. A vérifier en préprod ?

pierre-24 commented:
Poke @Alex-D : pour la "cassure" du design, tu saurais jeter un coup d'oeil ? Je plaide coupable dans tout les cas, mais je vois pas à quoi toucher dans le CSS (donc dans le doute, j'ai touché à rien)

Pour les licences, comme ça fait pareil sur les tutos, [j'ai posé la question d'abord](http://zestedesavoir.com/forums/sujet/668/enlever-la-licence-ne-fonctionne-pas/).

Alex-D commented on templates/article/member/view_online.html:
```diff
@@ -53,9 +59,6 @@ <h1 {% if article.image %}class="illu"{% endif %}>
     {{ article.txt|safe }}
-
-
-    {% include "article/includes/pager.part.html" %}
```
Pourquoi tu as supprimé ça ? Il ne faut pas y toucher, surtout que ça ne concerne pas les licences. C'est hors PR + pas bien :-°

pierre-24 replied:
Parce que ça y est pas en _off_ et que ça n'as pas de sens. Ceci dit, je suis d'accord que c'est hors de portée de la PR (et je viens de comprendre sont utilité, donc je fait un _rewind_ là dessus).

Alex-D commented:
>  je vois pas à quoi toucher dans le CSS

Y a pas à toucher au CSS en fait, tout est déjà fait, les tutos et les articles ont la même structure. Si tu as fait pareil, ça doit fonctionner. Ca se peut que tu n'ais pas le CSS à jour par contre.

---

Je pense que @firm1 va pas être content qu'aucun test n'ait été rédigé :-°

firm1 commented:
En effet, un petit test ne serait pas de trop par ici (surtout que la structure interne des articles est impactée).

Je rajoute cependant une note pour la QA : pensez à vérifier que la PR ne casse pas les articles existants en prod. Pour cela, il suffit de crée un article avant cette PR, le publier, et ensuite merger la PR, refaire la QA de retrocompatibilité.

pierre-24 commented:
> Y a pas à toucher au CSS en fait, tout est déjà fait, les tutos et les articles ont la même structure. Si tu as fait pareil, ça doit fonctionner. Ca se peut que tu n'ais pas le CSS à jour par contre.

J'ai le CSS de la branche master, c'est pas le plus récent ?

> En effet, un petit test ne serait pas de trop par ici (surtout que la structure interne des articles est impactée).

Je crois pas qu'il y aie _factory_ pour les licences, mais je vais chercher, ou à défaut en créer une.

> Je rajoute cependant une note pour la QA : pensez à vérifier que la PR ne casse pas les articles existants en prod. Pour cela, il suffit de crée un article avant cette PR, le publier, et ensuite merger la PR, refaire la QA de retrocompatibilité.

Merci pour le truc :)

pierre-24 committed:
Retour du pager sur view_online.html

(j'avais pas suivi son utilité, merci Alex-D)

Eskimon commented:
Je referais une QA complète ce soir si les modifs sont faites entre temps :)
Pour le CSS, pour ma part c'est clair il est pas a jour. Si de dist/ peut être uploade comme il faut ce serait super (ignorez ce message si c'est déjà le cas)

pierre-24 commented:
> Je referais une QA complète ce soir si les modifs sont faites entre temps :)

[Tant que je sais pas quel est le bon comportement sur les licences de tutos](http://zestedesavoir.com/forums/sujet/668/enlever-la-licence-ne-fonctionne-pas/), je peux pas le copier pour les licences d'articles :p (mais d'ici là, j'aurais de toute façon écrit le test unitaire).

pierre-24 commented:
Pour @Alex-D : 

sans licence : 

![Sans-licence](http://zestedesavoir.com/media/galleries/407/7f515c26-8bb5-4b93-b692-b27cb0889a02.png.960x960_q85.png)

et avec : 

![avec-licence](http://zestedesavoir.com/media/galleries/407/331a90f6-5944-44a6-88f5-8d6bb2d27b74.png.960x960_q85.png)

... J'ai le css de la branche "master" il me semble, regénéré via GULP et tout ça.

pierre-24 committed:
Ajout test unitaire et autorise licence nulle

Suite à la QA d'Eskimon :
- Il est maintenant possible d'enlever la licence dans le
  formulaire de modification
- Un test unitaire est ajouté

pierre-24 commented:
Ping @Eskimon : t'avais promis que tu ferais de la QA si c'était bon. Pour moi c'est bon, j'ai même testé avec un article créé via une autre branche (ce qui dispense pas de le tenter quand même)

poulp commented on zds/article/tests.py:
```diff
@@ -371,6 +375,134 @@ def test_url_for_staff(self):
+        # logout before
+        self.client.logout()
+        # login with author
+        self.assertEqual(
```
self.assertTrue(...)

pierre-24 replied:
Bien vu. Je ne savais pas que ça existait :)

pierre-24 committed:
changer les assertEqual() en assertTrue()

pour les `self.client.login()`

pierre-24 committed:
Ajout d'un test réglant une erreur sous windows

artragis commented:
It works for me. ^^ sympa la PR.

Eskimon commented:
Je merge dès que cette histoire de CSS est réglée (ping @Alex-D ou @sandhose )

Alex-D commented:
Le style est bien celui attendu, il n'y a rien à toucher, comme je l'avais déjà dit en fait :D

pierre-24 commented:
Merge, _then_ :)

Eskimon commented:
Ok, a monitorer pour le rendu quand meme ca me parait bizarre st'histoire (que les tags soient decolle de la barre de titre ca fait un peu lait non ?)

pierre-24 commented:
> Ok, a monitorer pour le rendu quand meme ca me parait bizarre st'histoire (que les tags soient decolle de la barre de titre ca fait un peu lait non ?)

Faudra checker ça en prod' avec le bon CSS :)

Alex-D commented:
> que les tags soient decolle de la barre de titre ca fait un peu lait non ?

Quand j'avais testé ça fonctionnait bien. Je verrais au pire, c'est du détail, et ça casse rien.
